File: frontend/src/physics/modes/mesh3d.ts
import type { PhysicsEngine } from "../PhysicsEngine";
import { ParticleSystem } from "../ParticleSystem";

export const mesh3dMode = {
  name: "mesh3d" as const,
  dim: 3 as const,

  defaultParams(): Record<string, number> {
    return {
      grid_size: 5,
      spacing: 60,
      stiffness: 20,
      spring_damping: 3,
      mass: 1.5,
      max_speed: 200,
      particle_mass: 1.5,
      particle_radius_factor: 5,
      bounds_margin: 250,
      drag: 2,
      gravity_x: 0,
      gravity_y: 0,
      gravity_z: 0,
    };
  },

  setup(engine: PhysicsEngine, params: Record<string, number>): void {
    const p = { ...this.defaultParams(), ...params };
    engine.reset();
    engine.dim = 3;
    engine.gravity = new Float32Array([p.gravity_x, p.gravity_y, p.gravity_z]);
    engine.dragCoefficient = p.drag;
    engine.maxSpeed = p.max_speed;
    engine.enableCollisions = true;

    const gs = Math.floor(p.grid_size);
    const spacing = p.spacing;
    const mass = p.particle_mass;
    const margin = p.bounds_margin;
    const k = p.stiffness;
    const kd = p.spring_damping;
    const diag = spacing * Math.SQRT2;

    const positions: number[][] = [];
    for (let x = 0; x < gs; x++) {
      for (let y = 0; y < gs; y++) {
        for (let z = 0; z < gs; z++) {
          positions.push([
            (x - gs / 2) * spacing,
            (y - gs / 2) * spacing,
            (z - gs / 2) * spacing,
          ]);
        }
      }
    }

    const n = positions.length;
    const masses = new Array(n).fill(mass);
    engine.particles = ParticleSystem.create(positions, masses);

    // Small random initial velocities
    const vel = engine.particles.velocities;
    for (let i = 0; i < n * 3; i++) {
      vel[i] = (Math.random() - 0.5) * 2.0;
    }

    // Collision radii
    engine.collisionRadii = new Float32Array(n);
    engine.collisionRadii.fill(mass * p.particle_radius_factor);

    // Bounds
    engine.bounds = {
      minPos: new Float32Array([-margin, -margin, -margin]),
      maxPos: new Float32Array([margin, margin, margin]),
    };
    engine.boundsMode = "elastic";

    const idx = (x: number, y: number, z: number) => x * gs * gs + y * gs + z;

    // Interior structural + diagonal springs
    for (let x = 0; x < gs - 1; x++) {
      for (let y = 0; y < gs - 1; y++) {
        for (let z = 0; z < gs - 1; z++) {
          engine.springs.add(idx(x, y, z), idx(x + 1, y, z), spacing, k, kd);
          engine.springs.add(idx(x, y, z), idx(x, y + 1, z), spacing, k, kd);
          engine.springs.add(idx(x, y, z), idx(x, y, z + 1), spacing, k, kd);
          engine.springs.add(idx(x, y, z), idx(x + 1, y + 1, z), diag, k, kd);
          engine.springs.add(idx(x, y + 1, z), idx(x + 1, y, z), diag, k, kd);
          engine.springs.add(idx(x, y, z), idx(x + 1, y, z + 1), diag, k, kd);
          engine.springs.add(idx(x, y, z + 1), idx(x + 1, y, z), diag, k, kd);
          engine.springs.add(idx(x, y, z), idx(x, y + 1, z + 1), diag, k, kd);
          engine.springs.add(idx(x, y, z + 1), idx(x, y + 1, z), diag, k, kd);
        }
      }
    }

    // Edge springs for faces not covered by interior loop
    for (let y = 0; y < gs - 1; y++) {
      for (let z = 0; z < gs - 1; z++) {
        engine.springs.add(idx(gs - 1, y, z), idx(gs - 1, y + 1, z), spacing, k, kd);
        engine.springs.add(idx(gs - 1, y, z), idx(gs - 1, y, z + 1), spacing, k, kd);
      }
    }
    for (let x = 0; x < gs - 1; x++) {
      for (let z = 0; z < gs - 1; z++) {
        engine.springs.add(idx(x, gs - 1, z), idx(x + 1, gs - 1, z), spacing, k, kd);
        engine.springs.add(idx(x, gs - 1, z), idx(x, gs - 1, z + 1), spacing, k, kd);
      }
    }
    for (let x = 0; x < gs - 1; x++) {
      for (let y = 0; y < gs - 1; y++) {
        engine.springs.add(idx(x, y, gs - 1), idx(x + 1, y, gs - 1), spacing, k, kd);
        engine.springs.add(idx(x, y, gs - 1), idx(x, y + 1, gs - 1), spacing, k, kd);
      }
    }
    for (let x = 0; x < gs - 1; x++) {
      engine.springs.add(idx(x, gs - 1, gs - 1), idx(x + 1, gs - 1, gs - 1), spacing, k, kd);
    }
    for (let y = 0; y < gs - 1; y++) {
      engine.springs.add(idx(gs - 1, y, gs - 1), idx(gs - 1, y + 1, gs - 1), spacing, k, kd);
    }
    for (let z = 0; z < gs - 1; z++) {
      engine.springs.add(idx(gs - 1, gs - 1, z), idx(gs - 1, gs - 1, z + 1), spacing, k, kd);
    }
  },
};

File: backend/modes/mesh3d.py
from __future__ import annotations
import math
import torch
from .base import SimulationMode
from backend.physics.engine import PhysicsEngine
from backend.physics.particle import ParticleSystem
from backend.physics.collisions import Bounds


class Mesh3DMode(SimulationMode):
    """3D particle mesh grid with force-based springs, elastic collisions,
    and boundary reflection. Matches the gorillasun spring model."""

    name = "mesh3d"
    dim = 3

    def default_params(self) -> dict:
        return {
            "grid_size": 5,
            "spacing": 60.0,
            "stiffness": 20.0,
            "spring_damping": 3.0,
            "mass": 1.5,
            "max_speed": 200.0,
            "particle_mass": 1.5,
            "particle_radius_factor": 5.0,
            "bounds_margin": 250.0,
            "drag": 2.0,
            "gravity_x": 0.0,
            "gravity_y": 0.0,
            "gravity_z": 0.0,
        }

    def setup(self, engine: PhysicsEngine, params: dict) -> None:
        p = {**self.default_params(), **params}
        engine.reset()
        engine.dim = self.dim
        engine.gravity = torch.tensor([p["gravity_x"], p["gravity_y"], p["gravity_z"]])
        engine.drag_coefficient = p["drag"]
        engine.max_speed = p["max_speed"]
        engine.enable_collisions = True

        gs = int(p["grid_size"])
        spacing = p["spacing"]
        mass = p["particle_mass"]
        margin = p["bounds_margin"]
        k = p["stiffness"]
        kd = p["spring_damping"]
        diag = spacing * math.sqrt(2)

        # Build 3D particle grid
        positions = []
        for x in range(gs):
            for y in range(gs):
                for z in range(gs):
                    px = (x - gs / 2) * spacing
                    py = (y - gs / 2) * spacing
                    pz = (z - gs / 2) * spacing
                    positions.append([px, py, pz])

        n = len(positions)
        engine.particles = ParticleSystem.create(
            positions=positions,
            masses=[mass] * n,
        )

        # Give particles small random initial velocities
        engine.particles.velocities = (torch.rand(n, 3) - 0.5) * 2.0

        # Collision radii
        engine.collision_radii = torch.full((n,), mass * p["particle_radius_factor"])

        # Bounds
        engine.bounds = Bounds(
            min_pos=torch.tensor([-margin, -margin, -margin]),
            max_pos=torch.tensor([margin, margin, margin]),
        )
        engine.bounds_mode = "elastic"

        # Helper to get flat index from 3D grid coords
        def idx(x: int, y: int, z: int) -> int:
            return x * gs * gs + y * gs + z

        # Build springs matching mesh.html connectivity
        # Interior cells: structural + diagonal springs
        for x in range(gs - 1):
            for y in range(gs - 1):
                for z in range(gs - 1):
                    # Structural (axis-aligned)
                    engine.springs.add(idx(x, y, z), idx(x + 1, y, z), spacing, k, kd)
                    engine.springs.add(idx(x, y, z), idx(x, y + 1, z), spacing, k, kd)
                    engine.springs.add(idx(x, y, z), idx(x, y, z + 1), spacing, k, kd)
                    # Shear/diagonal springs
                    engine.springs.add(idx(x, y, z), idx(x + 1, y + 1, z), diag, k, kd)
                    engine.springs.add(idx(x, y + 1, z), idx(x + 1, y, z), diag, k, kd)
                    engine.springs.add(idx(x, y, z), idx(x + 1, y, z + 1), diag, k, kd)
                    engine.springs.add(idx(x, y, z + 1), idx(x + 1, y, z), diag, k, kd)
                    engine.springs.add(idx(x, y, z), idx(x, y + 1, z + 1), diag, k, kd)
                    engine.springs.add(idx(x, y, z + 1), idx(x, y + 1, z), diag, k, kd)

        # Edge springs for faces not covered by interior loop
        for y in range(gs - 1):
            for z in range(gs - 1):
                engine.springs.add(idx(gs - 1, y, z), idx(gs - 1, y + 1, z), spacing, k, kd)
                engine.springs.add(idx(gs - 1, y, z), idx(gs - 1, y, z + 1), spacing, k, kd)

        for x in range(gs - 1):
            for z in range(gs - 1):
                engine.springs.add(idx(x, gs - 1, z), idx(x + 1, gs - 1, z), spacing, k, kd)
                engine.springs.add(idx(x, gs - 1, z), idx(x, gs - 1, z + 1), spacing, k, kd)

        for x in range(gs - 1):
            for y in range(gs - 1):
                engine.springs.add(idx(x, y, gs - 1), idx(x + 1, y, gs - 1), spacing, k, kd)
                engine.springs.add(idx(x, y, gs - 1), idx(x, y + 1, gs - 1), spacing, k, kd)

        for x in range(gs - 1):
            engine.springs.add(idx(x, gs - 1, gs - 1), idx(x + 1, gs - 1, gs - 1), spacing, k, kd)
        for y in range(gs - 1):
            engine.springs.add(idx(gs - 1, y, gs - 1), idx(gs - 1, y + 1, gs - 1), spacing, k, kd)
        for z in range(gs - 1):
            engine.springs.add(idx(gs - 1, gs - 1, z), idx(gs - 1, gs - 1, z + 1), spacing, k, kd)
